include/oneapi/dpl/internal/scan_by_segment_impl.h

```diff
@@ -0,0 +1,374 @@
+                                          __identity, __binary_op, __wgroup_size); // need to use exclusive scan delta
+
+                    // 1c. Update local partial reductions and write to global memory.
+                    for (::std::size_t __i = __start; __i < __end; ++__i)
+                    {
+                        __out_values[__i] = __binary_op(__carry_in, __out_values[__i]);
+
+                        if (__i >= __n - 1 || !__binary_pred(__keys[__i], __keys[__i + 1]))
+                            break;
+                    }
+
+                    if (__local_id == __wgroup_size - 1) // last work item writes the group's carry out
+                    {
+                        ::std::size_t __group_id = __item.get_group(0);
+
+                        __seg_ends_acc[__group_id] = __group_has_segment_break;
+
+                        if (__max_end == NO_SEGMENT_BREAK)
+                        {
+                            __partials_acc[__group_id] = __binary_op(__carry_in, __accumulator);
+                        }
+                        else
+                        {
+                            __partials_acc[__group_id] = __accumulator;
+                        }
+                    }
+                });
+        });
+
+        // 2. Apply work group carry outs, calculate output indices, and load results into correct indices.
+        __exec.queue()
+            .submit([&](sycl::handler& __cgh) {
+                oneapi::dpl::__ranges::__require_access(__cgh, __keys, __out_values);
+
+                auto __partials_acc = __partials.template get_access<sycl::access_mode::read>(__cgh);
+                auto __seg_ends_acc = __seg_ends.template get_access<sycl::access_mode::read>(__cgh);
+
+                __cgh.depends_on(__wg_scan);
+
+                __dpl_sycl::__local_accessor<__val_type> __loc_partials_acc(__wgroup_size, __cgh);
+
+                __dpl_sycl::__local_accessor<bool> __loc_seg_ends_acc(__wgroup_size, __cgh);
+#if _ONEDPL_COMPILE_KERNEL && _ONEDPL_KERNEL_BUNDLE_PRESENT
+                __cgh.use_kernel_bundle(__kernel2.get_kernel_bundle());
+#endif
+                __cgh.parallel_for<_SegScanKernel2>(
+#if _ONEDPL_COMPILE_KERNEL && !_ONEDPL_KERNEL_BUNDLE_PRESENT
+                    __kernel2,
+#endif
+                    sycl::nd_range<1>{__n_groups * __wgroup_size, __wgroup_size}, [=](sycl::nd_item<1> __item) {
+                        auto __group = __item.get_group();
+                        ::std::size_t __group_id = __item.get_group(0);
+                        ::std::size_t __global_id = __item.get_global_id(0);
+                        ::std::size_t __local_id = __item.get_local_id(0);
+                        ::std::size_t __start = __global_id * __vals_per_item;
+                        ::std::size_t __end = __dpl_sycl::__minimum<decltype(__n)>{}(__start + __vals_per_item, __n);
+
+                        ::std::int32_t __wg_agg_idx = __group_id - 1;
```

Comment:
is `std::int32t` sufficient? are we fine with potentially negative values?

Reply:
We do not apply any aggregates with `__group_id` being 0, so the negative is disregarded in this case. The signed type is necessary to prevent an unsigned overflow below 0 for the arithmetic in the loop below.

I believe `int32_t` is sufficient since it stores the group ID meaning this would only cause an issue when we have over 2 billion+ work groups.